From c14d64971e5a69f80906651a85518f0ed604f415 Mon Sep 17 00:00:00 2001
From: Jianwei Fan <jianwei.fan@rock-chips.com>
Date: Mon, 13 May 2024 17:12:51 +0800
Subject: [PATCH 288/305] media: i2c: rk628: fix problem of loop pulling down
 and up cec hpd

Type: Fix
Redmine ID: N/A
Associated modifications: N/A
Test: N/A

Change-Id: I4fdb6d95269b836d664e3530424f1aaf54fb0a64
Signed-off-by: Jianwei Fan <jianwei.fan@rock-chips.com>
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628_bt1120_v4l2.c |  2 +-
 drivers/media/i2c/rk628/rk628_csi_v4l2.c    |  2 +-
 drivers/media/i2c/rk628/rk628_hdmirx.c      | 24 ++++++++++++++++++++-
 drivers/media/i2c/rk628/rk628_hdmirx.h      |  2 ++
 4 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
index 2ba973cc5cc981..12898620cf8d90 100644
--- a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
@@ -539,7 +539,7 @@ static void rk628_hdmirx_hpd_ctrl(struct v4l2_subdev *sd, bool en)
 			HOT_PLUG_DETECT_MASK, HOT_PLUG_DETECT(set_level));
 
 	if (bt1120->cec_enable && bt1120->cec)
-		rk628_hdmirx_cec_hpd(bt1120->cec, en);
+		rk628_hdmirx_cec_hpd(bt1120->cec, tx_5v_power_present(sd));
 }
 
 static int rk628_bt1120_s_ctrl_detect_tx_5v(struct v4l2_subdev *sd)
diff --git a/drivers/media/i2c/rk628/rk628_csi_v4l2.c b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
index e47233de04a7a9..5974f0af85ab1b 100644
--- a/drivers/media/i2c/rk628/rk628_csi_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
@@ -604,7 +604,7 @@ static void rk628_hdmirx_hpd_ctrl(struct v4l2_subdev *sd, bool en)
 			HOT_PLUG_DETECT_MASK, HOT_PLUG_DETECT(set_level));
 
 	if (csi->cec_enable && csi->cec)
-		rk628_hdmirx_cec_hpd(csi->cec, en);
+		rk628_hdmirx_cec_hpd(csi->cec, tx_5v_power_present(sd));
 }
 
 
diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.c b/drivers/media/i2c/rk628/rk628_hdmirx.c
index bb4329d8106289..20269efe550dd9 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.c
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.c
@@ -975,6 +975,17 @@ void rk628_hdmirx_cec_irq(struct rk628 *rk628, struct rk628_hdmirx_cec *cec)
 }
 EXPORT_SYMBOL(rk628_hdmirx_cec_irq);
 
+static void rk628_delayed_work_cec(struct work_struct *work)
+{
+	struct delayed_work *dwork = to_delayed_work(work);
+	struct rk628_hdmirx_cec *cec = container_of(dwork, struct rk628_hdmirx_cec,
+						    delayed_work_cec);
+	bool en = rk628_hdmirx_tx_5v_power_detect(cec->rk628->hdmirx_det_gpio);
+
+	cec->cec_hpd = en;
+	cec_queue_pin_hpd_event(cec->adap, en, ktime_get());
+}
+
 struct rk628_hdmirx_cec *rk628_hdmirx_cec_register(struct rk628 *rk628)
 {
 	struct rk628_hdmirx_cec *cec;
@@ -1016,6 +1027,8 @@ struct rk628_hdmirx_cec *rk628_hdmirx_cec_register(struct rk628 *rk628)
 	/* override the module pointer */
 	cec->adap->owner = THIS_MODULE;
 
+	INIT_DELAYED_WORK(&cec->delayed_work_cec, rk628_delayed_work_cec);
+
 	ret = devm_add_action(cec->dev, rk628_hdmirx_cec_del, cec);
 	if (ret) {
 		cec_delete_adapter(cec->adap);
@@ -1058,6 +1071,8 @@ struct rk628_hdmirx_cec *rk628_hdmirx_cec_register(struct rk628 *rk628)
 	 */
 	devm_remove_action(cec->dev, rk628_hdmirx_cec_del, cec);
 
+	schedule_delayed_work(&cec->delayed_work_cec, msecs_to_jiffies(10000));
+
 	return cec;
 }
 EXPORT_SYMBOL(rk628_hdmirx_cec_register);
@@ -1081,7 +1096,11 @@ void rk628_hdmirx_cec_hpd(struct rk628_hdmirx_cec *cec, bool en)
 	if (!cec || !cec->adap)
 		return;
 
-	cec_queue_pin_hpd_event(cec->adap, en, ktime_get());
+	rk628_dbg(cec->rk628, "%s: cec_hpd:%d, en:%d\n", __func__, cec->cec_hpd, en);
+	if (cec->cec_hpd != en) {
+		cec->cec_hpd = en;
+		cec_queue_pin_hpd_event(cec->adap, en, ktime_get());
+	}
 }
 EXPORT_SYMBOL(rk628_hdmirx_cec_hpd);
 
@@ -1091,6 +1110,9 @@ void rk628_hdmirx_cec_state_reconfiguration(struct rk628 *rk628,
 	unsigned int irqs;
 	u32 val;
 
+	/* clk_hdmirx_cec = 32.768k */
+	rk628_clk_set_rate(rk628, CGU_CLK_HDMIRX_CEC, 32768);
+
 	rk628_i2c_write(rk628, HDMI_RX_CEC_ADDR_L, cec->addresses & 0xff);
 	rk628_i2c_write(rk628, HDMI_RX_CEC_ADDR_H, (cec->addresses >> 8) & 0xff);
 
diff --git a/drivers/media/i2c/rk628/rk628_hdmirx.h b/drivers/media/i2c/rk628/rk628_hdmirx.h
index 98dd68ffe8d748..9e7f9b12922fd9 100644
--- a/drivers/media/i2c/rk628/rk628_hdmirx.h
+++ b/drivers/media/i2c/rk628/rk628_hdmirx.h
@@ -503,7 +503,9 @@ struct rk628_hdmirx_cec {
 	unsigned int tx_status;
 	bool tx_done;
 	bool rx_done;
+	bool cec_hpd;
 	struct cec_notifier *notify;
+	struct delayed_work delayed_work_cec;
 };
 
 void rk628_hdmirx_set_hdcp(struct rk628 *rk628, struct rk628_hdcp *hdcp, bool en);
-- 
2.17.1

